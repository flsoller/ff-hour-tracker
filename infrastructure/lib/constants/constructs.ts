export const USERS_SERVICE = {
  NAME: "hour-tracker-users-service",
  LAMBDA_INTEGRATION: "hour-tracker-users-service-lambda-integration",
};

export const API_GATEWAY = {
  NAME: "hour-tracker-api-gateway",
};

export const DEFAULT_AUTHORIZER = {
  NAME: "hour-tracker-default-authorizer-service",
  LAMBDA_AUTHORIZER: "hour-tracker-default-lambda-authorizer",
};

export const AUTHENTICATOR_SERVICE = {
  NAME: "hour-tracker-authentication-service",
  LAMBDA_INTEGRATION: "hour-tracker-authentication-service-lambda-integration",
};

export const ORGANIZATION_MANAGER = {
  NAME: "hour-tracker-organization-manager",
};
