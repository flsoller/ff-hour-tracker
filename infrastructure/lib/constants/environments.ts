export const ENVIRONMENT_PARAMS = {
  DATABASE_URL_KEY: "HOUR_TRACKER_DATABASE_URL",
  JWT_SECRET: "HOUR_TRACKER_JWT_SECRET",
  API_DOMAIN_NAME: "HOUR_TRACKER_API_DOMAIN_NAME",
  API_DOMAIN_CERT_ARN: "HOUR_TRACKER_API_DOMAIN_CERT_ARN",
};
