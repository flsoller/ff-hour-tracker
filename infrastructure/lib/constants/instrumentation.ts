export const DEFAULT_INSTRUMENTATION_CONFIG = {
  NEW_RELIC_EXTENSION_SEND_FUNCTION_LOGS: "true",
  NEW_RELIC_LAMBDA_EXTENSION_ENABLED: "true",
  NEW_RELIC_DATA_COLLECTION_TIMEOUT: "10s",
  NEW_RELIC_IGNORE_EXTENSION_CHECKS: "all",
  NEW_RELIC_NO_CONFIG_FILE: "true",
};
