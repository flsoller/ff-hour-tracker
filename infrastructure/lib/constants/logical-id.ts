const AppName = "HourTrackerAPI";

export const LOGICAL_ID = {
  HOUR_TRACKER_API_GATEWAY: `${AppName}-Gateway`,
  HOUR_TRACKER_API_AUTHENTICATOR: `${AppName}-Authenticator`,
  HOUR_TRACKER_API_AUTHORIZER: `${AppName}-Authorizer`,
  HOUR_TRACKER_API_MEMBERS: `${AppName}-Members`,
  HOUR_TRACKER_API_ORG_MANAGER: `${AppName}-OrgManager`,
};
