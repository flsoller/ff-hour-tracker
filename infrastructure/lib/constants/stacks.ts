export const HOUR_TRACKER = {
  API_AUTHORIZER: "hour-tracker-api-authorizer",
  API_AUTHENTICATOR: "hour-tracker-api-authenticator",
  API_GATEWAY: "hour-tracker-api-gatewayv2",
  API_MEMBERS: "hour-tracker-api-members",
};
