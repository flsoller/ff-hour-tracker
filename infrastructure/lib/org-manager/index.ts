import * as cdk from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Duration } from "aws-cdk-lib";
import { Construct } from "constructs";
import { RemovalPolicy } from "aws-cdk-lib";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../constants/ecr";
import { ORGANIZATION_MANAGER } from "../constants/constructs";
import { ENVIRONMENT_PARAMS } from "../constants/environments";
import { LOGICAL_ID } from "../constants/logical-id";

export class HourTrackerOrganizationManager extends cdk.Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const dbConnectionString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.DATABASE_URL_KEY
    );
    const supportUserEmail = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.SUPPORT_USER_EMAIL
    );
    const supportUserPassword = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.SUPPORT_USER_PW
    );

    const newRelicAccountId = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_ACCOUNT_ID
    );

    const newRelicIngestLicense = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_LICENSE_KEY
    );

    const hashOrVersion = new cdk.CfnParameter(this, "hashOrVersion", {
      description: "The version or hash used to reference the images",
      type: "String",
    }).valueAsString;

    const ecrRepo = ecr.Repository.fromRepositoryName(
      this,
      `${id}-repo`,
      HOUR_TRACKER_ECR_REPO_NAMES.ORGANIZATION_MANAGER
    );

    const organizationManager = new lambda.DockerImageFunction(
      this,
      ORGANIZATION_MANAGER.NAME,
      {
        code: lambda.DockerImageCode.fromEcr(ecrRepo, {
          tagOrDigest: hashOrVersion,
        }),
        architecture: lambda.Architecture.ARM_64,
        environment: {
          DATABASE_URL: dbConnectionString,
          SUPPORT_USER_EMAIL: supportUserEmail,
          SUPPORT_USER_PW: supportUserPassword,
          NEW_RELIC_LAMBDA_HANDLER: "apps/org-manager/index.handler",
          NEW_RELIC_ACCOUNT_ID: newRelicAccountId,
          NEW_RELIC_LICENSE_KEY: newRelicIngestLicense,
        },
        timeout: Duration.seconds(20),
        functionName: LOGICAL_ID.HOUR_TRACKER_API_ORG_MANAGER,
      }
    );
    organizationManager.applyRemovalPolicy(RemovalPolicy.DESTROY);
  }
}
