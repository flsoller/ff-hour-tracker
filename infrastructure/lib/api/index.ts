import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { AuthorizerService } from "./authorizer";
import { HourTrackerApiGateway } from "./gateway";
import { HOUR_TRACKER } from "../constants/stacks";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { ENVIRONMENT_PARAMS } from "../constants/environments";
import { AuthenticationService } from "./authenticator";
import { MembersService } from "./members";

export class HourTrackerApi extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const hashOrVersion = new cdk.CfnParameter(this, "hashOrVersion", {
      description: "The version or hash used to reference the images",
      type: "String",
    }).valueAsString;

    const dbConnectionString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.DATABASE_URL_KEY
    );

    const jwtSecretString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.JWT_SECRET
    );

    const defaultAuthorizer = new AuthorizerService(
      this,
      HOUR_TRACKER.API_AUTHORIZER,
      {
        hashOrVersion,
        dbConnectionString,
        jwtSecretString,
      }
    );

    const apiGateway = new HourTrackerApiGateway(
      this,
      HOUR_TRACKER.API_GATEWAY,
      {
        authService: defaultAuthorizer.authorizer,
      }
    );

    new AuthenticationService(this, HOUR_TRACKER.API_AUTHENTICATOR, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
      jwtSecretString,
    });

    new MembersService(this, HOUR_TRACKER.API_MEMBERS, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
    });
  }
}
