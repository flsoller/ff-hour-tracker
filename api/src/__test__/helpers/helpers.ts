import { prisma } from '../../utils/prisma';
import { registerUser } from '../../services/auth';
import { IRegisterUser } from '@hour-tracker/core-types/api/auth';
import supertest from 'supertest';
import { Role } from '@prisma/client';

async function createOrganizations() {
  const organization = await prisma.organization.create({
    data: {
      name: 'organization',
      description: 'organization description',
    },
  });

  const otherOrganization = await prisma.organization.create({
    data: {
      name: 'otherOrganization',
      description: 'otherOrganization description',
    },
  });

  return [organization, otherOrganization];
}

async function createMemberForOrganization(
  orgId: string,
  emailAddress = 'test@mail.com',
  firstName = 'Test',
  lastName = 'User',
) {
  return prisma.member.create({
    data: {
      firstName,
      lastName,
      emailAddress,
      orgId,
    },
  });
}

async function createUserForOrganization(orgId: string, options = {}) {
  const params: IRegisterUser = {
    emailAddress: 'testuser@db.com',
    password: '12345',
    name: 'Mr. Test',
    orgId,
    role: Role.USER,
    ...options,
  };

  return registerUser(params);
}

async function loginTestUser(
  app: Express.Application,
  emailAddress: string,
): Promise<string> {
  return supertest(app)
    .post('/api/v0/auth/signin')
    .send({
      emailAddress,
      password: '12345',
    })
    .then((res) => res.body.accessToken);
}

export {
  createOrganizations,
  createMemberForOrganization,
  createUserForOrganization,
  loginTestUser,
};
