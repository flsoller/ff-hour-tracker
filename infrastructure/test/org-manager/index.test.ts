import { Template } from "aws-cdk-lib/assertions";
import * as cdk from "aws-cdk-lib";
import { cdkResourceFinderWithTypeFilter } from "../helpers";
import { HourTrackerOrganizationManager } from "../../lib/org-manager";
import { ORGANIZATION_MANAGER } from "../../lib/constants/constructs";

let app: cdk.App;
let orgManagerStack: HourTrackerOrganizationManager;
let template: Template;

beforeEach(() => {
  app = new cdk.App({
    context: { "@aws-cdk/core:newStyleStackSynthesis": false },
  });
  orgManagerStack = new HourTrackerOrganizationManager(
    app,
    "HourTrackerOrganizationManager"
  );
  template = Template.fromStack(orgManagerStack);
});

describe("HourTrackerOrganizationManager", () => {
  it("should contain the correct Lambda properties", () => {
    const { DependsOn, ...attributesToCheck } = cdkResourceFinderWithTypeFilter(
      template,
      "AWS::Lambda::Function",
      ORGANIZATION_MANAGER.NAME
    );
    expect(attributesToCheck).toMatchSnapshot();
  });
});
