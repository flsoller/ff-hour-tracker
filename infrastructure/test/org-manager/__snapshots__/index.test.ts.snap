// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourTrackerOrganizationManager should contain the correct Lambda properties 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "Architectures": [
      "arm64",
    ],
    "Code": {
      "ImageUri": {
        "Fn::Join": [
          "",
          [
            {
              "Ref": "AWS::AccountId",
            },
            ".dkr.ecr.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/hour-tracker-org-manager:",
            {
              "Ref": "hashOrVersion",
            },
          ],
        ],
      },
    },
    "Environment": {
      "Variables": {
        "DATABASE_URL": {
          "Ref": "SsmParameterValueHOURTRACKERDATABASEURLC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_ACCOUNT_ID": {
          "Ref": "SsmParameterValueNEWRELICACCOUNTIDC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_DATA_COLLECTION_TIMEOUT": "10s",
        "NEW_RELIC_EXTENSION_SEND_FUNCTION_LOGS": "true",
        "NEW_RELIC_IGNORE_EXTENSION_CHECKS": "all",
        "NEW_RELIC_LAMBDA_EXTENSION_ENABLED": "true",
        "NEW_RELIC_LAMBDA_HANDLER": "apps/org-manager/index.handler",
        "NEW_RELIC_LICENSE_KEY": {
          "Ref": "SsmParameterValueNEWRELICLICENSEKEYC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "NEW_RELIC_NO_CONFIG_FILE": "true",
        "SUPPORT_USER_EMAIL": {
          "Ref": "SsmParameterValueSUPPORTUSEREMAILC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "SUPPORT_USER_PW": {
          "Ref": "SsmParameterValueSUPPORTUSERPWC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
      },
    },
    "FunctionName": "HourTrackerAPI-OrgManager",
    "PackageType": "Image",
    "Role": {
      "Fn::GetAtt": [
        "hourtrackerorganizationmanagerServiceRoleCE3CF813",
        "Arn",
      ],
    },
    "Timeout": 20,
  },
  "Type": "AWS::Lambda::Function",
  "UpdateReplacePolicy": "Delete",
}
`;
